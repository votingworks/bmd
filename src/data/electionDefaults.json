{
  "bmdConfig": {
    "requireActivation": true,
    "showSettingsPage": true
  }
}
