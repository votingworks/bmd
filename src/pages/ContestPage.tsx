import React, { useContext, useEffect, useState } from 'react'
import { RouteComponentProps } from 'react-router-dom'
import { CandidateVote, OptionalYesNoVote } from '@votingworks/ballot-encoder'

import ordinal from '../utils/ordinal'

import BallotContext from '../contexts/ballotContext'

import CandidateContest from '../components/CandidateContest'
import ElectionInfo from '../components/ElectionInfo'
import LinkButton from '../components/LinkButton'
import Prose from '../components/Prose'
import Screen from '../components/Screen'
import Sidebar from '../components/Sidebar'
import Text from '../components/Text'
import YesNoContest from '../components/YesNoContest'
import SettingsTextSize from '../components/SettingsTextSize'
import TextIcon from '../components/TextIcon'
import MsEitherNeitherContest from '../components/MsEitherNeitherContest'

interface ContestParams {
  contestNumber: string
}

const ContestPage = (props: RouteComponentProps<ContestParams>) => {
  const isReviewMode = window.location.hash === '#review'
  const { contestNumber } = props.match.params
  const {
    ballotStyleId,
    contests,
    election,
    precinctId,
    setUserSettings,
    updateVote,
    userSettings,
    votes,
  } = useContext(BallotContext)
  const currentContestIndex = parseInt(contestNumber, 10)
  const contest = contests[currentContestIndex]

  const vote = votes[contest.id]

  const [isVoteComplete, setIsVoteComplete] = useState(false)

  const prevContestIndex = currentContestIndex - 1
  const prevContest = contests[prevContestIndex]

  const nextContestIndex = currentContestIndex + 1
  const nextContest = contests[nextContestIndex]

  useEffect(() => {
    const calculateIsVoteComplete = () => {
      if (contest.type === 'yesno') {
        setIsVoteComplete(!!vote)
        return
      }
      if (contest.type === 'candidate') {
        setIsVoteComplete(
          contest.seats === ((vote as CandidateVote) ?? []).length
        )
        return
      }
      if (contest.type === 'ms-either-neither') {
        setIsVoteComplete(
          votes[contest.pickOneContestId]?.length === 1 ||
            votes[contest.eitherNeitherContestId]?.[0] === 'no'
        )
        return
      }
    }
    calculateIsVoteComplete()
  }, [contest, vote, votes])

  return (
    <Screen>
      {contest.type === 'candidate' && (
        <CandidateContest
          aria-live="assertive"
          key={contest.id}
          contest={contest}
          parties={election.parties}
          vote={(vote ?? []) as CandidateVote}
          updateVote={updateVote}
        />
      )}
      {contest.type === 'yesno' && (
        <YesNoContest
          key={contest.id}
          contest={contest}
          vote={vote as OptionalYesNoVote}
          updateVote={updateVote}
        />
      )}
      {contest.type === 'ms-either-neither' && (
        <MsEitherNeitherContest
          key={contest.id}
          contest={contest}
          eitherNeitherContestVote={
            votes[contest.eitherNeitherContestId] as OptionalYesNoVote
          }
          pickOneContestVote={
            votes[contest.pickOneContestId] as OptionalYesNoVote
          }
          updateVote={updateVote}
        />
      )}
      <Sidebar
        footer={
          <React.Fragment>
            <SettingsTextSize
              userSettings={userSettings}
              setUserSettings={setUserSettings}
            />
            <ElectionInfo
              election={election}
              ballotStyleId={ballotStyleId}
              precinctId={precinctId}
              horizontal
            />
          </React.Fragment>
        }
      >
        <Prose>
          <Text center>
            This is the <strong>{ordinal(currentContestIndex + 1)}</strong> of{' '}
            {contests.length} contests.
          </Text>
          {isReviewMode ? (
            <React.Fragment>
              <p>
                <LinkButton
                  big
                  primary={isVoteComplete}
                  to={`/review#contest-${contest.id}`}
                  id="next"
                >
                  <TextIcon arrowRight white={isVoteComplete}>
                    Review
                  </TextIcon>
                </LinkButton>
              </p>
            </React.Fragment>
          ) : (
            <React.Fragment>
              <p>
                <LinkButton
                  big
                  id="next"
                  primary={isVoteComplete}
                  to={nextContest ? `/contests/${nextContestIndex}` : '/review'}
                >
                  <TextIcon arrowRight white={isVoteComplete}>
                    Next
                  </TextIcon>
                </LinkButton>
              </p>
              <p>
                <LinkButton
                  small
                  id="previous"
                  to={prevContest ? `/contests/${prevContestIndex}` : '/'}
                >
                  <TextIcon arrowLeft>Back</TextIcon>
                </LinkButton>
              </p>
            </React.Fragment>
          )}
        </Prose>
      </Sidebar>
    </Screen>
  )
}

export default ContestPage
