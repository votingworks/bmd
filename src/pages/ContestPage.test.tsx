import React from 'react'
import { Route } from 'react-router-dom'

import { render } from '../../test/testUtils'

import electionDefaults from '../data/electionDefaults.json'
import electionSample from '../data/electionSample.json'

const firstContestTitle = electionSample.contests[0].title

import ContestPage from './ContestPage'

it(`renders ContestPage`, () => {
  const { container, getByText } = render(
    <Route path="/contests/:contestNumber" component={ContestPage} />,
    {
      route: `/contests/0`,
    }
  )
  getByText(firstContestTitle)
  expect(container).toMatchSnapshot()
})

it(`redirects to top-level page if no contest number match`, () => {
  const resetBallot = jest.fn()
  const homeMock = () => <div>Home Mock</div>
  const { getByText } = render(
    <>
      <Route path="/contests/:contestNumber" component={ContestPage} />
      <Route exact path="/" render={homeMock} />
    </>,
    {
      resetBallot,
      route: '/contests/666666',
    }
  )

  expect(resetBallot).toBeCalled()
  getByText('Home Mock')
})

it(`doesn't display help and settings pages if disabled`, () => {
  const { queryByText } = render(
    <Route path="/contests/:contestNumber" component={ContestPage} />,
    {
      election: {
        ...electionDefaults,
        ...electionSample,
        ...{
          bmdConfig: {
            showHelpPage: false,
            showSettingsPage: false,
          },
        },
      },
      route: `/contests/0`,
    }
  )
  expect(queryByText('Help')).toBeFalsy()
  expect(queryByText('Settings')).toBeFalsy()
})
