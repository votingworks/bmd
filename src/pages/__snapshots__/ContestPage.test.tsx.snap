// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders ContestPage 1`] = `
.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap-reverse;
  -ms-flex-wrap: wrap-reverse;
  flex-wrap: wrap-reverse;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  border-bottom: 1px solid rgb(169,169,169);
  background: #455a64;
  padding: 0.25rem;
}

.c11 > *:first-child {
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
}

.c11 > * {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0.25rem;
}

.c14 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-wrap: wrap-reverse;
  -ms-flex-wrap: wrap-reverse;
  flex-wrap: wrap-reverse;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-order: -1;
  -ms-flex-order: -1;
  order: -1;
  border-bottom: 1px solid rgb(169,169,169);
  background: #455a64;
  padding: 0.25rem;
}

.c14 > *:first-child {
  -webkit-order: 2;
  -ms-flex-order: 2;
  order: 2;
}

.c14 > * {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0.25rem;
}

.c12 {
  border: none;
  border-radius: 0.25rem;
  box-sizing: border-box;
  background: rgb(211,211,211);
  cursor: pointer;
  padding: 0.75rem 1rem;
  line-height: 1.25;
  color: black;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  width: 100%;
  max-width: 66ch;
  line-height: 1.2;
}

.c2 h1 {
  margin: 2rem 0 1rem;
  line-height: 1.1;
  font-size: 1.5rem;
}

.c2 h2 {
  margin: 1.5rem 0 0.75rem;
  font-size: 1.25rem;
}

.c2 h3,
.c2 p {
  margin-top: 1rem;
  margin-bottom: 1rem;
  font-size: 1rem;
}

.c2 h1 + h2 {
  margin-top: -0.75rem;
}

.c2 h1 + p,
.c2 h2 + p {
  margin-top: -0.75rem;
}

.c2 h3 + p {
  margin-top: -1rem;
}

.c2 > :first-child {
  margin-top: 0;
}

.c2 > :last-child {
  margin-bottom: 0;
}

.c10 {
  word-break: break-word;
}

.c1 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 1rem 0.75rem 0.5rem;
}

.c3 {
  text-transform: uppercase;
  font-size: 0.85rem;
  font-weight: 600;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
  overflow: auto;
}

.c4::before,
.c4::after {
  position: absolute;
  z-index: 1;
  width: 100%;
  height: 0.25rem;
  content: '';
  -webkit-transition: opacity 0.25s ease;
  transition: opacity 0.25s ease;
}

.c4::before {
  top: 0;
  opacity: 0;
  background: linear-gradient( to bottom, rgb(177,186,190) 0%, transparent 100% );
}

.c4::after {
  bottom: 0;
  opacity: 0;
  background: linear-gradient( to bottom, transparent 0%, rgb(177,186,190) 100% );
}

.c5 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  overflow: auto;
}

.c6 {
  margin: 0 auto;
  width: 100%;
  max-width: 35rem;
  padding: 0.5rem 0.5rem 1rem;
}

.c7 {
  display: grid;
  grid-auto-rows: minmax(auto,1fr);
  grid-gap: 0.75rem;
}

.c8 {
  display: grid;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  border-radius: 0.125rem;
  box-shadow: 0 0.125rem 0.125rem 0 rgba(0,0,0,0.14), 0 0.1875rem 0.0625rem -0.125rem rgba(0,0,0,0.12), 0 0.0625rem 0.3125rem 0 rgba(0,0,0,0.2);
  background: #FFFFFF;
  cursor: pointer;
  -webkit-transition: background 0.25s,color 0.25s;
  transition: background 0.25s,color 0.25s;
}

button.c8 {
  text-align: left;
}

.c8:focus-within {
  outline: rgb(77,144,254) dashed 0.25rem;
}

.c8::before {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  border-right: 1px solid;
  border-color: rgb(211,211,211);
  border-radius: 0.125rem 0 0 0.125rem;
  background: #FFFFFF;
  width: 3rem;
  text-align: center;
  color: #028099;
  font-size: 2rem;
  font-weight: 700;
  content: '';
}

.c8 > div {
  padding: 0.5rem 0.5rem 0.5rem 4rem;
}

.c9 {
  margin-right: 0.5rem;
}

.c13 {
  text-align: center;
  color: #FFFFFF;
  -webkit-flex: 2;
  -ms-flex: 2;
  flex: 2;
}

@media (min-width:480px) {
  .c11 > *:only-child {
    -webkit-flex: 0;
    -ms-flex: 0;
    flex: 0;
    margin: auto;
    min-width: 33.333%;
  }
}

@media print {
  .c11 {
    display: none;
  }
}

@media (min-width:480px) {
  .c14 > *:first-child {
    margin-left: auto;
  }
}

@media (min-width:480px) {
  .c14 > * {
    -webkit-flex: 0;
    -ms-flex: 0;
    flex: 0;
  }
}

@media (min-width:480px) {
  .c14 > *:only-child {
    -webkit-flex: 0;
    -ms-flex: 0;
    flex: 0;
    min-width: 33.333%;
  }
}

@media print {
  .c14 {
    display: none;
  }
}

@media print {
  .c0 {
    -webkit-box-pack: start;
    -webkit-justify-content: flex-start;
    -ms-flex-pack: start;
    justify-content: flex-start;
    padding: 0;
  }
}

@media (min-width:480px) {
  .c2 {
    line-height: 1.3;
  }
}

@media (min-width:720px) {
  .c1 {
    padding: 1rem 1.5rem 0.5rem;
    padding-left: 5rem;
  }
}

@media (min-width:720px) {
  .c6 {
    padding-right: 1rem;
    padding-left: 1rem;
  }
}

@media (min-width:480px) {
  .c8 > div {
    padding: 1rem 1rem 1rem 4rem;
  }
}

@media print {
  .c13 {
    color: #FFFFFF;
  }
}

<div>
  <main
    class="c0"
  >
    <div
      class="c1"
      id="contest-header"
    >
      <div
        class="c2"
        id="audiofocus"
      >
        <h1
          aria-label="President and Vice-President."
        >
          <div
            class="c3"
          >
            United States
          </div>
          President and Vice-President
        </h1>
        <p
          aria-label="Vote for 1. You have selected 0. Use the down arrow to hear your options. Use the right arrow to move to the next contest."
        >
          <strong>
            Vote for 
            1
            .
          </strong>
           You have selected
           
          0
          .
        </p>
      </div>
    </div>
    <div
      class="c4"
    >
      <div
        class="c5"
      >
        <div
          class="c6"
        >
          <div
            class="c7"
          >
            <label
              aria-label="Joseph Barchi and Joseph Hallaren, Federalist."
              class="c8"
              for="barchi-hallaren"
            >
              <input
                class="visually-hidden c9"
                id="barchi-hallaren"
                name="Joseph Barchi and Joseph Hallaren"
                role="option"
                type="checkbox"
                value="barchi-hallaren"
              />
              <div
                class="c2"
              >
                <p
                  class="c10"
                >
                  <strong>
                    Joseph Barchi and Joseph Hallaren
                  </strong>
                  <br />
                  Federalist
                </p>
              </div>
            </label>
            <label
              aria-label="Adam Cramer and Greg Vuocolo, People's."
              class="c8"
              for="cramer-vuocolo"
            >
              <input
                class="visually-hidden c9"
                id="cramer-vuocolo"
                name="Adam Cramer and Greg Vuocolo"
                role="option"
                type="checkbox"
                value="cramer-vuocolo"
              />
              <div
                class="c2"
              >
                <p
                  class="c10"
                >
                  <strong>
                    Adam Cramer and Greg Vuocolo
                  </strong>
                  <br />
                  People's
                </p>
              </div>
            </label>
            <label
              aria-label="Daniel Court and Amy Blumhardt, Liberty."
              class="c8"
              for="court-blumhardt"
            >
              <input
                class="visually-hidden c9"
                id="court-blumhardt"
                name="Daniel Court and Amy Blumhardt"
                role="option"
                type="checkbox"
                value="court-blumhardt"
              />
              <div
                class="c2"
              >
                <p
                  class="c10"
                >
                  <strong>
                    Daniel Court and Amy Blumhardt
                  </strong>
                  <br />
                  Liberty
                </p>
              </div>
            </label>
            <label
              aria-label="Alvin Boone and James Lian, Constitution."
              class="c8"
              for="boone-lian"
            >
              <input
                class="visually-hidden c9"
                id="boone-lian"
                name="Alvin Boone and James Lian"
                role="option"
                type="checkbox"
                value="boone-lian"
              />
              <div
                class="c2"
              >
                <p
                  class="c10"
                >
                  <strong>
                    Alvin Boone and James Lian
                  </strong>
                  <br />
                  Constitution
                </p>
              </div>
            </label>
            <label
              aria-label="Ashley Hildebrand-McDougall and James Garritty, Whig."
              class="c8"
              for="hildebrand-garritty"
            >
              <input
                class="visually-hidden c9"
                id="hildebrand-garritty"
                name="Ashley Hildebrand-McDougall and James Garritty"
                role="option"
                type="checkbox"
                value="hildebrand-garritty"
              />
              <div
                class="c2"
              >
                <p
                  class="c10"
                >
                  <strong>
                    Ashley Hildebrand-McDougall and James Garritty
                  </strong>
                  <br />
                  Whig
                </p>
              </div>
            </label>
            <label
              aria-label="Martin Patterson and Clay Lariviere, Labor."
              class="c8"
              for="patterson-lariviere"
            >
              <input
                class="visually-hidden c9"
                id="patterson-lariviere"
                name="Martin Patterson and Clay Lariviere"
                role="option"
                type="checkbox"
                value="patterson-lariviere"
              />
              <div
                class="c2"
              >
                <p
                  class="c10"
                >
                  <strong>
                    Martin Patterson and Clay Lariviere
                  </strong>
                  <br />
                  Labor
                </p>
              </div>
            </label>
          </div>
        </div>
      </div>
    </div>
  </main>
  <nav
    class="c11"
  >
    <button
      class="c12"
      id="next"
      role="option"
      type="button"
    >
      Next
    </button>
    <button
      class="c12"
      id="previous"
      role="option"
      type="button"
    >
      Back
    </button>
    <p
      class="c13"
    >
      1
       of 
      22
    </p>
  </nav>
  <nav
    class="c14"
  >
    <div />
    <button
      class="c12"
      role="option"
      type="button"
    >
      Settings
    </button>
  </nav>
</div>
`;
